File: ml-controller/services/allocator_contract_guard.py
"""Allocator-contract runtime guards for replay and dry-run jobs.

The guard is intentionally separate from the allocator itself. It only blocks
side effects and constrains replay dates when an operator explicitly enables
allocator contract verification.
"""
from __future__ import annotations

import logging
import os

logger = logging.getLogger(__name__)

ALLOCATOR_CONTRACT_GUARD_ENV = "STOCKVISION_ALLOCATOR_CONTRACT_GUARD"
ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE_ENV = "STOCKVISION_ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE"


def env_truthy(name: str) -> bool:
    return os.environ.get(name, "").strip().lower() in {"1", "true", "yes", "on"}


def allocator_contract_guard_enabled() -> bool:
    return env_truthy(ALLOCATOR_CONTRACT_GUARD_ENV)


def assert_allocator_contract_run_date(run_date: str, *, label: str) -> None:
    """Fail closed so allocator contract runs cannot replay arbitrary dates."""
    if not allocator_contract_guard_enabled():
        return
    allowed = os.environ.get(ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE_ENV, "").strip()
    if not allowed:
        raise RuntimeError(
            f"{label}: {ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE_ENV} is required when "
            f"{ALLOCATOR_CONTRACT_GUARD_ENV}=1"
        )
    if not run_date:
        raise RuntimeError(f"{label}: explicit run_date is required for allocator contract guard")
    if run_date != allowed:
        raise RuntimeError(
            f"{label}: allocator contract run_date={run_date} blocked; allowed={allowed}"
        )
    logger.warning(
        "[AllocatorContractGuard] %s enabled for run_date=%s; callbacks and writes must be no-op",
        label,
        run_date,
    )

File: worker/src/node-runner/allocatorContractGuard.ts
export const ALLOCATOR_CONTRACT_GUARD_ENV = 'STOCKVISION_ALLOCATOR_CONTRACT_GUARD'
export const ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE_ENV = 'STOCKVISION_ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE'

export function envTruthy(name: string): boolean {
  return ['1', 'true', 'yes', 'on'].includes((process.env[name] ?? '').trim().toLowerCase())
}

export function allocatorContractGuardEnabled(): boolean {
  return envTruthy(ALLOCATOR_CONTRACT_GUARD_ENV)
}

export function assertAllocatorContractRunDate(runDate: string, label: string): void {
  if (!allocatorContractGuardEnabled()) return
  const allowed = (process.env[ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE_ENV] ?? '').trim()
  if (!allowed) {
    throw new Error(`${ALLOCATOR_CONTRACT_ALLOWED_RUN_DATE_ENV} is required when ${ALLOCATOR_CONTRACT_GUARD_ENV}=1`)
  }
  if (!runDate) throw new Error(`${label}: explicit run_date is required for allocator contract guard`)
  if (runDate !== allowed) throw new Error(`${label}: allocator contract run_date=${runDate} blocked; allowed=${allowed}`)
  console.warn(`[AllocatorContractGuard] ${label} enabled for run_date=${runDate}; D1/KV writes are no-op`)
}
